### DISCUSSION TITLE
Deprecate `{is_ajax_request}`?

### DISCUSSION THREAD

**Discussion by robsonsobral** [ExpressionEngine/ExpressionEngine#3213] [Feature Requests]
Now, the Frontend developer in me will fight the backend developers in all of you.

---

The conditional `{is_ajax_request}` depends on `X-Requested-With` header to work. However, `X-Requested-With` isn't a valid header, neither is submitted using the standard [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API), but something some libs used to add when using `XMLHttpRequest()`.[^1]

I gonna add instructions on how to make `AJAX_REQUEST`[^2] constant to work on up-to-date browsers and modern JavaScript to the `{is_ajax_request}` documentation (for now).

[^1]: There's a myth backend devs are still in love with jQuery, landlines and bank agencies, which will all disappear. Maybe the myth is real.
[^2]: Who still says "AJAX" these days? Do they never heard of JSON?

**Comment by intoeetive** [comment]
I can see the pain on having to add `X-Requested-With` header if you're making AJAX requests with vanilla JS, or a framework that does not set it internally.

However, even though `fetch()` method has been added to JS, it does not provide reliable way to identify itself as such for the backend - the data are still being sent as regular HTTP POST/GET request.

Unless I am missing something, the only way to identify whether request was done via AJAX call (whether `XMLHttpRequest` or `fetch`) is to check the headers. We could also check for `'Content-Type': 'application/json'` but I'm not sure if that would also give some false positives.

I'd love to hear more thoughts on how we can make it better!

@Yulyaswan feel free to join the discussion if you want

**Comment by robsonsobral** [comment]
Yeah. The intention of the specification is to not give this kind of clue.

My suggestion would be:

- allow [templates of different types to have the same name](https://github.com/ExpressionEngine/ExpressionEngine/issues/1143);
- add `.json`~~, `.xml`[^1]~~ and `.txt` as accepted template types;
- chose the template to be parsed using the `Accept` header, **not** the `Content-type`;

[^1]: I never understood the reason for the `.feed` extension;
